components/director/pkg/webhook/types.go

```diff
@@ -0,0 +1,259 @@
+	}
+
+	return nil
+}
+
+func (ot *Response) Validate() error {
+	if ot.Location == nil {
+		return errors.New("missing Output Template location field")
+	}
+
+	if ot.SuccessStatusCode == nil {
+		return errors.New("missing Output Template success status code field")
+	}
+
+	if ot.Error == nil {
+		return errors.New("missing Output Template error field")
+	}
+
+	return nil
+}
+
+func (st *ResponseStatus) Validate() error {
+	if st.Status == nil {
+		return errors.New("missing Status Template status field")
+	}
+
+	if st.SuccessStatusCode == nil {
+		return errors.New("missing Status Template success status code field")
+	}
+
+	if st.SuccessStatusIdentifier == nil {
+		return errors.New("missing Status Template success status identifier field")
+	}
+
+	if st.InProgressStatusIdentifier == nil {
+		return errors.New("missing Status Template in progress status identifier field")
+	}
+
+	if st.FailedStatusIdentifier == nil {
+		return errors.New("missing Status Template failed status identifier field")
+	}
+
+	if st.Error == nil {
+		return errors.New("missing Status Template error field")
+	}
+
+	return nil
+}
+
+func ParseURLTemplate(tmpl *string, reqData RequestData) (*URL, error) {
+	if tmpl == nil {
+		return nil, nil
+	}
+
+	urlTemplate, err := template.New("url").Option("missingkey=zero").Parse(*tmpl)
+	if err != nil {
+		return nil, err
+	}
+
+	result := new(bytes.Buffer)
+	err = urlTemplate.Execute(result, reqData)
+	if err != nil {
+		return nil, err
+	}
+
+	var url URL
+	if err := json.Unmarshal(result.Bytes(), &url); err != nil {
+		return nil, err
+	}
+
+	return &url, url.Validate()
+}
+
+func ParseInputTemplate(tmpl *string, reqData RequestData) ([]byte, error) {
+	if tmpl == nil {
+		return nil, nil
+	}
+
+	inputTemplate, err := template.New("input").Option("missingkey=zero").Parse(*tmpl)
+	if err != nil {
+		return nil, err
+	}
+
+	result := new(bytes.Buffer)
+	if err := inputTemplate.Execute(result, reqData); err != nil {
+		return nil, err
+	}
+
+	res := json.RawMessage{}
+	if err := json.Unmarshal(result.Bytes(), &res); err != nil {
+		return nil, err
+	}
+
+	return result.Bytes(), nil
+}
+
+func ParseHeadersTemplate(tmpl *string, reqData RequestData) (http.Header, error) {
+	if tmpl == nil {
+		return nil, nil
+	}
+
+	headerTemplate, err := template.New("header").Option("missingkey=zero").Parse(*tmpl)
+	if err != nil {
+		return nil, err
+	}
+
+	result := new(bytes.Buffer)
+	if err := headerTemplate.Execute(result, reqData); err != nil {
+		return nil, err
+	}
+
+	if result.Len() == 0 {
+		return http.Header{}, nil
+	}
+
+	var headers http.Header
+	if err := json.Unmarshal(result.Bytes(), &headers); err != nil {
+		return nil, err
+	}
+
+	return headers, nil
+}
+
+func ParseOutputTemplate(inputTmpl, outputTmpl *string, respData ResponseData) (*Response, error) {
+	if outputTmpl == nil && inputTmpl != nil {
+		return nil, errors.New("missing webhook output template")
+	}
+
+	if outputTmpl == nil {
```

Comment:
such behaviour should be avoided, this is hidden/unexpected and will often result in bugs - e.g. whoever calls Parseoutput template should rather get an error that he provided an invalid param and not call it rather than get a nil,nil